src/sema/ast.rs

```diff
@@ -1526,18 +1526,21 @@ impl Statement {
     pub fn reachable(&self) -> bool {
         match self {
             Statement::Block { statements, .. } => statements.iter().all(|s| s.reachable()),
-            Statement::Underscore(_) | Statement::Destructure(..) | Statement::VariableDecl(..) => {
-                true
-            }
+            Statement::Underscore(_)
+            | Statement::Destructure(..)
+            | Statement::VariableDecl(..)
+            | Statement::Emit { .. }
+            | Statement::Delete(..) => true,
+
+            Statement::Continue(_) | Statement::Break(_) | Statement::Return(..) => false,
+
             Statement::If(_, reachable, ..)
             | Statement::While(_, reachable, ..)
             | Statement::DoWhile(_, reachable, ..)
-            | Statement::Expression(_, reachable, _) => *reachable,
-            Statement::Emit { .. } => true,
-            Statement::Delete(..) => true,
-            Statement::Continue(_) | Statement::Break(_) | Statement::Return(..) => false,
-            Statement::For { reachable, .. } | Statement::TryCatch(_, reachable, _) => *reachable,
-            Statement::Assembly(_) => unimplemented!("Assembly block ast not ready"),
```

Comment:
@seanyoung This PR remove the `unimplemented!` from here. I haven't added any test.

Reply:
Why no test?

Reply:
Presently there is no test for the `reachable` function. I can create a unity test and test it altogether. Should I do so?

Reply:
There are tests for unreachable statements: tests/contract_testcases/substrate/functions/for_forever.sol and tests/contract_testcases/substrate/loops/test_infinite_loop.sol

That's not great coverage, but best to improve this by adding tests for solidity statements after assembly statements.

Reply:
I've included a contract test case and a unity test.